src/dataflow/src/render/reduce.rs

```diff
@@ -628,32 +626,12 @@ impl KeyValPlan {
             .map(aggregates.iter().map(|a| a.expr.clone()))
             .project(input_arity..(input_arity + aggregates.len()));
 
-        // Determine the columns we'll need from the row.
-        let mut demand = Vec::new();
-        demand.extend(key_mfp.demand());
-        demand.extend(val_mfp.demand());
-        demand.sort();
-        demand.dedup();
-        // remap column references to the subset we use.
-        let mut demand_map = std::collections::HashMap::new();
-        for column in demand.iter() {
-            demand_map.insert(*column, demand_map.len());
-        }
-        key_mfp.permute(demand_map.clone(), demand_map.len());
         key_mfp.optimize();
         let key_plan = key_mfp.into_plan().unwrap().into_nontemporal().unwrap();
-        let demand_map_len = demand_map.len();
-        val_mfp.permute(demand_map, demand_map_len);
         val_mfp.optimize();
```

Comment:
Please check that moving the place of the `optimize` calls to `key_mfp` and `val_mfp` do not change their outcome. At the moment, the MFP is permuted based on the demand map. After this PR lands, the optimization is applied before the demand analysis. @frankmcsherry

Reply:
I think this is probably ok, from a correctness point of view. The `demand` stuff is analogous to `permute`, in that it is trying to think about which columns we want to pluck out of the `Row` we are presented with. Tbh, I think it's value is now diminished with the `Permutation` stuff, as we are decoding all `Datum`s rather than just those identified by `demand`. Perhaps in the future, we redo the `demand` into `skips` thing here for all uses of `Row` data.

At the moment, things seem a bit inconsistent, as we determine `demand` and `skips`, but then decode and de-permute everything, and then subset out the elements we want. It might make sense to wipe it for the moment? It is probably not actively harmful, other than being complexity without attendant benefit at the moment.

Reply:
> At the moment, things seem a bit inconsistent, as we determine demand and skips, but then decode and de-permute everything

I don't understand what you mean. Aren't we indeed only decoding the stuff asked for by `skips` , or did I misunderstand this code?

```
                let mut row_datums = row_datums.borrow_with_many(row_parts);

                let mut row_iter = row_datums.drain(..);
                let mut datums_local = datums.borrow();
                // Unpack only the demanded columns.
                for skip in skips.iter() {
                    datums_local.push((&mut row_iter).nth(*skip).unwrap());
                }
 ```

Reply:
Anyway, if both of you think this is OK, I'll leave it as-is for now in the interest of getting this PR landed, and someone who understands this code better (possibly including future me) can clean it up.